src/main/scala/akka/persistence/cassandra/snapshot/CassandraSnapshotStore.scala

```diff
@@ -166,49 +166,43 @@ class CassandraSnapshotStore(cfg: Config) extends SnapshotStore with CassandraSt
   }
 
   private def metadata(prepStmt: PreparedStatement, persistenceId: String, criteria: SnapshotSelectionCriteria): Future[Vector[SnapshotMetadata]] = {
-    // TODO the RowIterator is using some blocking, would benefit from a rewrite
-    Future {
-      new RowIterator(prepStmt, persistenceId, criteria.maxSequenceNr).map { row =>
-        SnapshotMetadata(row.getString("persistence_id"), row.getLong("sequence_nr"), row.getLong("timestamp"))
-      }.dropWhile(_.timestamp > criteria.maxTimestamp).toVector
-    }(blockingDispatcher)
-  }
+    def iterate(nr: Long, buf: ListBuffer[SnapshotMetadata]): Future[Vector[SnapshotMetadata]] = {
+      session.execute(prepStmt.bind(persistenceId, nr: JLong)) flatMap { rs =>
+        var rowCount = 0
+        var nextNr = 0l
+
+        def fetchAvailable(rs: ResultSet): Future[Vector[SnapshotMetadata]] = {
+          for (r <- rs.asScala.take(rs.getAvailableWithoutFetching)) {
+            nextNr = r.getLong("sequence_nr")
+            rowCount += 1
+            buf += SnapshotMetadata(r.getString("persistence_id"), nextNr, r.getLong("timestamp"))
+          }
+          if (rowCount < maxMetadataResultSize) Future.successful(buf.dropWhile(_.timestamp > criteria.maxTimestamp).toVector)
+          else if (rs.isFullyFetched) iterate(nextNr - 1, buf)
+          else rs.fetchMoreResults().asScala.flatMap(fetchAvailable)
+        }
 
-  private class RowIterator(prepStmt: PreparedStatement, persistenceId: String, maxSequenceNr: Long) extends Iterator[Row] {
-    var currentSequenceNr = maxSequenceNr
-    var rowCount = 0
-
-    // we know that the session is initialized, since we got the prepStmt
-    val ses = Await.result(session.underlying(), config.blockingTimeout)
-
-    // FIXME more blocking
-    var iter = newIter()
-    def newIter() = ses.execute(prepStmt.bind(persistenceId, currentSequenceNr: JLong)).iterator
-
-    @annotation.tailrec
-    final def hasNext: Boolean =
-      if (iter.hasNext)
-        true
-      else if (rowCount < maxMetadataResultSize)
-        false
-      else {
-        rowCount = 0
-        currentSequenceNr -= 1
-        iter = newIter()
-        hasNext
+        fetchAvailable(rs)
       }
-
-    def next(): Row = {
-      val row = iter.next()
-      currentSequenceNr = row.getLong("sequence_nr")
-      rowCount += 1
-      row
     }
+    iterate(criteria.maxSequenceNr, ListBuffer.empty)
   }
 
 }
 
 private[snapshot] object CassandraSnapshotStore {
+
+  implicit class ListenableFutureOps[V](lf: ListenableFuture[V]) {
+    def asScala = {
+      val p = Promise[V]()
+      Futures.addCallback(lf, new FutureCallback[V] {
+        override def onFailure(t: Throwable): Unit = p.failure(t)
+        override def onSuccess(result: V): Unit = p.success(result)
+      })
+      p.future
+    }
+  }
```

Comment:
note that there is a similar (but more implicit) conversion in `package.scala`
This looks better, and I would prefer the `asScala` approach everywhere. As a first step you can perhaps place this in the  `package.scala` alongside the other conversion.